Convert  phpUnit tests to Kahlan for UseAtmon class

MarikaBBB opened the pull request:
This Pr converts the existing phpUnit tests from the UseAtom class to Kahlan. I followed the Kahlan documentation to structure the tests into clear sections using ```describe```, ```it``` , and ```beforeEach``` blocks.

Key changes:
- Implemented test for ```::registered()``` to verify that actions ```init``` and ```wp_head``` are registered with the correct callbacks using ```\WP_Mock```
- Converted test for ```::init()``` to manage filter and action
- Converted test for  ```::wpHead()``` to ensure correct ink generation
- I used```beforeEach``` and ```afterEach``` for setup and teardown.
- I used ```\WP_Mock``` in the same way as in the original phpUnit test file.
- Faced difficulties with ```expect($results)->toBe()``` syntax initially but I resolved by checking Kahlan matcher methods.

MarikaBBB commented:
This Pr converts the existing phpUnit tests from the UseAtom class to Kahlan. I followed the Kahlan documentation to structure the tests into clear sections using ```describe```, ```it``` , and ```beforeEach``` blocks.

Covere tests
- Action registration using ```::register()```
- Managing filter and action in ```::init()```
- Generating correct link output in ```::wpHead()```
- I used ```\WP_Mock``` in the same way as in the original phpUnit test file.

snim2 commented on composer.json:
```diff
@@ -6,7 +6,8 @@
     "require-dev": {
         "phpunit/phpunit": "^9.0",
```
So, at the end of this PR it would be good to have another commit to remove `phpunit` and `wp_mock` since by that time they won't be in use

MarikaBBB replied:
Thank you. I removed it and you can find the change in my last push.

snim2 commented on spec/UseAtomSpec.php:
```diff
@@ -14,13 +14,18 @@
-        \WP_Mock::expectActionAdded('wp_head', [$this->useAtom, 'wpHead']);
+    describe("::register()", function() {
+        it('should register actions correctly', function() {
+            // expect($this->UseAtom)->toBeAnInstanceOf(\Dxw\Iguana\Registerable::class);
```
Let's remove commented code

snim2 commented on spec/UseAtomSpec.php:
```diff
@@ -14,13 +14,18 @@
+    describe("::register()", function() {
+        it('should register actions correctly', function() {
+            // expect($this->UseAtom)->toBeAnInstanceOf(\Dxw\Iguana\Registerable::class);
+            \WP_Mock::expectActionAdded('init', [$this->useAtom, 'init']);
```
I'm not sure we need `WP_Mock` here for Kahlan tests, have a look at other repos where we use Kahlan and see if you can do the same thing with `allow()` which is more idiomatic in Kahlan.

snim2 reviewed (changes requested):
This is nearly there, but I think we can lose wp-mock. Also, when the Kahlan tests are sorted and the old dependencies removed, we should also change the GitHub action on this line:

https://github.com/MarikaBBB/iguana-extras/blob/main/.github/workflows/continuous-integration.yml#L30-L31

MarikaBBB commented:
This second part of PR is in relation the above suggested modifications:

What has been done:
- I removed wp-mock parts and rewrote them using ```allow()```  but made me not passing tests.

Next to do:
- remove ```phpunit``` and ```wp_mock``` in composer.json and in GitHub action

snim2 commented on spec/UseAtomSpec.php:
```diff
@@ -15,8 +15,9 @@
-            \WP_Mock::expectActionAdded('init', [$this->useAtom, 'init']);
-            \WP_Mock::expectActionAdded('wp_head', [$this->useAtom, 'wpHead']);
+            allow('add_action')->tobeCalled()->with('init', [$this->useAtom, 'init']);
+            allow('add_action')->tobeCalled()->with('wp_head', [$this->useAtom, 'wpHead']);
```
So you need some assertions here, and the `allow()` can be quite simple, because you want to test what is happening to `add_action` you don't want to mock it out completely. 

This test could be something like this:

```php
    describe("::register()", function() {
        it('should register actions correctly', function() {
            allow('add_action')->toBeCalled();
            expect('add_action')->toBeCalled()->times(2);
            expect('add_action')->toBeCalled()->once()->with('init', [$this->useAtom, 'init']);
            expect('add_action')->toBeCalled()->once()->with('wp_head', [$this->useAtom, 'wpHead']);

            $this->useAtom->register();
        });
    });
```

and also you have some other `WP_Mock` boilerplate in the file that needs to be removed

snim2 commented on spec/UseAtomSpec.php:
```diff
@@ -2,64 +2,82 @@
+			allow('remove_action')->toBeCalled();
+			expect('remove_action')->toBeCalled()->times(3);
+			expect('remove_action')->toBeCalled()->with('do_feed_rdf', 'do_feed_rdf', 10, 1);
+			allow('remove_action')->toBeCalled()->with('do_feed_rss', 'do_feed_rss', 10, 1);
```
I think these last two calls to `allow()` should be `expect()`?

MarikaBBB commented:
I implemented allow() and expect() as it should intended on the Kahlan documentation. 
I also changed name from ```tobeCalled()``` to ```toBeCalled()```

However, I encounter an error that do not let me pass one test.

Please see the follow:
```
UseAtom
  ::init()
    ✖ it adds the default_feed filter correctly
      an uncaught exception has been thrown in `vendor/kahlan/kahlan/src/Plugin/Monkey.php` line 87
      
      message:`Kahlan\PhpErrorException` Code(0) with message "`E_WARNING` call_user_func_array() expects parameter 1 to be a valid callback, function 'remove_action' not found or invalid function name"
```

snim2 commented on spec/UseAtomSpec.php:
```diff
@@ -2,64 +2,82 @@
+		$this->useAtom = new \Dxw\Iguana\Extras\UseAtom();
+
+        allow('add_filter')->toBeCalled();
+        allow('remove_action')->toBeCalled();
```
These calls to `allow()` should be in the relevant tests, not in the `beforeEach()` method

MarikaBBB replied:
I added to try to see if a different error would appear, but it did not. 
Now I have removed them. 
They are already in their relevant tests.
 I feel I have to write something else in the ```beforeEach()```, but I am unsure what.

MarikaBBB commented:
Successfully passed all tests.
I fixed the ```init()``` method.

The problem: 
- The test for the ```::init()``` method were failing because the ```add_filter``` and ```remove_action``` were not being triggered during the test execution

The why:
- the structure wasn't written accurately. I proceeded by groups ```allow()``` and ```expect()``` in two distinct groups. 
- I moved around ```$this->useAtom->ini()``` to see if there was a different error message. Then replaced again after the assertions block.

- I also added ```error_log()``` in the UseAtom.php for debugging

snim2 commented:
@MarikaBBB a couple of things to do next:

* You should also be able to remove `WP_Mock` from the composer files
* Run the lint over all the files:
```shell
❯ ./vendor/bin/php-cs-fixer fix --dry-run
PHP CS Fixer 3.68.0 Persian Successor by Fabien Potencier, Dariusz Ruminski and contributors.
PHP runtime: 8.3.15
Running analysis on 1 core sequentially.
You can enable parallel runner and speed up the analysis! Please see usage docs for more information.
Loaded config default from "/Users/sarah.mount/code/github.com/MarikaBBB/iguana-extras/.php-cs-fixer.php".
Using cache file ".php-cs-fixer.cache".
 3/3 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

   1) spec/UseAtomSpec.php
   2) src/UseAtom.php

Found 2 of 3 files that can be fixed in 0.027 seconds, 16.00 MB memory used
```
* Compare the old tests and the new and just check that they are testing the same thing

MarikaBBB commented:
I removed ```WP_Mock``` from the composer files
I compared old tests and new tests and find out I didn't implement the ```expect($this->useAtom)->toBeAnInstanceOf(\Dxw\Iguana\Registerable::class, $this->useAtom);``` in the ```::register()```
- I also polished the code by removing some lines

snim2 commented:
OK, I think the last thing is to deal with the GitHub Action file: https://github.com/MarikaBBB/iguana-extras/blob/main/.github/workflows/continuous-integration.yml
